rust/sdk/examples/chat/display.rs

```diff
@@ -127,6 +104,200 @@ pub async fn display(mut cell: ActoCell<Display, impl ActoRuntime>) {
+    identity: &Reader<Identity>,
+    text_area: &Reader<TextArea>,
+    messages: &Reader<Vec<Message>>,
+    scroll_state: &mut MessageScrollState,
+    not_connected: &Option<String>,
+) {
+    let is_editing = identity.project(|identity| identity.edit.is_some());
+    if is_editing {
+        render_editing_identity(f, identity);
+    } else {
+        render_chat(f, text_area, messages, scroll_state, not_connected);
+    }
+}
+
+fn render_editing_identity<W: io::Write>(f: &mut Frame<CrosstermBackend<W>>, identity: &Reader<Identity>) {
+    let size = f.size();
+    let pad = (size.height / 2).checked_sub(2).unwrap_or(0);
+    let layout = Layout::default()
+        .direction(ratatui::prelude::Direction::Vertical)
+        .constraints([
+            ratatui::prelude::Constraint::Length(pad),
+            ratatui::prelude::Constraint::Max(3),
+            ratatui::prelude::Constraint::Min(0),
+        ])
+        .split(size);
+
+    identity.project(|identity| {
+        if let Some(edit) = &identity.edit {
+            f.render_widget(edit.widget(), layout[1]);
+        }
+    });
+}
+
+fn render_chat<W: io::Write>(
+    f: &mut Frame<CrosstermBackend<W>>,
+    text_area: &Reader<TextArea>,
+    messages: &Reader<Vec<Message>>,
+    scroll_state: &mut MessageScrollState,
+    not_connected: &Option<String>,
+) {
+    let size = f.size();
+    let cmdheight = text_area.project(|t| t.lines().len() as u16);
+    let layout = Layout::default()
+        .direction(ratatui::prelude::Direction::Vertical)
+        .constraints([
+            ratatui::prelude::Constraint::Min(0),
+            ratatui::prelude::Constraint::Length(cmdheight + 2),
+        ])
+        .split(size);
+
+    messages.project(|msgs| {
+        let message_text_rect = layout[0];
+        let message_text_rect_width = message_text_rect.width;
+        let message_text_rect_height = message_text_rect.height;
+        tracing::info!("paragraph_available_width: {}", message_text_rect_width);
+        let mut lines = msgs
+            .into_iter()
+            .flat_map(|x| message_to_lines(x, message_text_rect_width))
+            .collect::<Vec<_>>();
+
+        // FIXME Maybe there is a better way to calculate height using Widget::render, Buffer, and binary search?
+        // That way using Paragraph::wrap can be possible
+        let text_height: u16 = lines.len().try_into().unwrap_or(0);
+
+        let scroll_range = text_height.checked_sub(message_text_rect_height).unwrap_or(0);
+        let user_scroll_y_from_bottom: u16 = {
+            // Is there a height shift from previous render?
+            let height_shift_on_update = if let Some(prev_height) = scroll_state.last_recorded_par_height {
+                text_height.checked_sub(prev_height).unwrap_or(0)
+            } else {
+                0
+            };
+
+            scroll_state.last_recorded_par_height = Some(text_height);
+
+            // if vscroll_from_bottom is not 0, it must shift up, because UX
```

Comment:
I think the code would become simpler by instead storing the offset from the bottom in the `scroll_state` and only converting that to the `scroll_y` when rendering the Paragraph. Growth of the Paragraph would have to increase the offset unless it is zero.

Reply:
That is now what the code does. Although it might have a little bit confusing

Reply:
Also, my comment is misleading

Reply:
should be clearer in https://github.com/Actyx/Actyx/pull/554/commits/a52d3233b2e80053dbe04d872393d0b23e2aa816

Reply:
yes, it is